WIP: diagnostics bundle

vemelin-epm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/kudobuilder/kudo/blob/master/CONTRIBUTING.md
2. Make sure you have added and ran the tests before submitting your PR
3. If the PR is unfinished, start it as a Draft PR: https://github.blog/2019-02-14-introducing-draft-pull-requests/
-->

**What this PR does / why we need it**:

**Important update**
this PR is declared overly complicated and investing into unnecessary extra flexibility and is therefore abandoned and kept for the reference.The simplified diagnostics in the first iteration is in progress here:
https://github.com/kudobuilder/kudo/pull/1506

<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #D2IQ-63414
- collect some instance resources (pods, replicasets etc.) and save as yml
- collect some kudo controller resources (pods, roles etc.) and save as yml
- "describe" for the resources mentioned above
- logs for instance and kudo controller

Still *WIP*,  things TODO:
tests, configuration, readiness to be run otherwise than as a command, possibly readiness to collect resources in parallel

**Problems**
- resources retrieval depends on other resources. I was struggling to provide an elegant way to exchange the data (normally, just names)
- resources should be placed in the order often opposite to the retrieval order, e.g. perhaps we may want pods under its owner, and OperatorVersion under Operator

**Current solution**
resource retrieval dependencies, directory hierarchy configuration are decoupled from resource collection logic. all resources are pre-collected before saving.

Example output of the current state (names and dirs configurable, pod logs are temporarily disabled, will be adapted to the new logic with the subsequent commits)
```
$ kubectl kudo diagnostics collect --instance instance-zk
diag
├── kudo
│   ├── ClusterRole_cluster-admin
│   │   └── cluster-admin.yml
│   ├── ServiceAccount_kudo-manager
│   │   └── kudo-manager.yml
│   ├── Service_kudo-controller-manager-service
│   │   └── kudo-controller-manager-service.yml
│   └── StatefulSet_kudo-controller-manager
│       ├── kudo-controller-manager.yml
│       └── Pod_kudo-controller-manager-0
│           └── kudo-controller-manager-0.yml
└── Operator_zookeeper
    ├── OperatorVersion_zookeeper-0.3.0
    │   ├── ClusterRole_cluster-admin
    │   │   └── cluster-admin.yml
    │   ├── Instance_zk
    │   │   └── zk.yml
    │   ├── ServiceAccount_default
    │   │   └── default.yml
    │   ├── Service_zk-cs
    │   │   └── zk-cs.yml
    │   ├── Service_zk-hs
    │   │   └── zk-hs.yml
    │   ├── StatefulSet_zk-zookeeper
    │   │   ├── Pod_zk-zookeeper-0
    │   │   │   └── zk-zookeeper-0.yml
    │   │   ├── Pod_zk-zookeeper-1
    │   │   │   └── zk-zookeeper-1.yml
    │   │   ├── Pod_zk-zookeeper-2
    │   │   │   └── zk-zookeeper-2.yml
    │   │   └── zk-zookeeper.yml
    │   └── zookeeper-0.3.0.yml
    └── zookeeper.yml
```

porridge commented on pkg/kudoctl/cmd/diagnostics/resource_providers.go:
```diff
@@ -0,0 +1,194 @@
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type resourceFuncs struct {
```
Can you please clarify what this is meant to abstract?

porridge commented on pkg/kudoctl/cmd/diagnostics/resources.go:
```diff
@@ -0,0 +1,226 @@
+
+type describeCollector struct {
+	config *restclient.Config
+	d      *resourceCollector
```
Why `d`?

porridge commented on pkg/kudoctl/cmd/diagnostics/resources.go:
```diff
@@ -0,0 +1,226 @@
+		Namespace: meta.GetNamespace(),
+		Name:      meta.GetName(),
+	}
+	w, err := f(info)
```
I know go style prefers short names but this borders on obfuscation :-)

porridge commented on pkg/kudoctl/cmd/diagnostics/scheme.go:
```diff
@@ -0,0 +1,24 @@
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// TODO: method copied from pkg/kudoctl/util/kudo. Probably should be made exported
```
:100: Please export, not copy.

porridge commented on pkg/kudoctl/cmd/diagnostics/writers.go:
```diff
@@ -0,0 +1,45 @@
+			return nil, err
+		}
+	}
+	return os.Create(dir + "/" + name)
```
please use `path.Join` here and elsewhere, this needs to work on Windows too I think

porridge commented on pkg/kudoctl/cmd/diagnostics/writers.go:
```diff
@@ -0,0 +1,45 @@
+		name = "errors.txt"
+	}
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		err = os.MkdirAll(dir, 0700)
```
I think you can run this unconditionally, it handles pre-existing `dir` just fine.

porridge commented on pkg/kudoctl/cmd/diagnostics/resources.go:
```diff
@@ -0,0 +1,226 @@
+package diagnostics
+
+/* collectors for kubernetes resources and related data, e.g logs and "describes" */
```
perhaps this file should be called `kubernetes.go`, then?

porridge commented on pkg/kudoctl/cmd/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,105 @@
+	ssc := resourceListCollector{getResources: cResources.statefulSets()}
+	rbc := resourceListCollector{getResources: cResources.roleBindings()}
+	crbc := resourceListCollector{getResources: cResources.clusterRoleBindings()}
+	rc := resourceListCollector{getResources: cResources.roles(&rbc)}
```
I looked at a couple of these functions and it does not look like they use the arguments.
If they are independent then you can do without intermediate variables and just replace this with a series of:
```
collectors = append(collectors, resourceListCollector{...})
```
or such.

zen-dog commented:
1. We need to simplify the diag file structure a bit. It's currently too verbose and has levels for no apparent benefit. I think most of the "non-nested" resources like `ClusterRole` do not need an enclosing folder:
Example:
```
diag
├── kudo
│   ├── ClusterRole_cluster-admin
│   │   └── cluster-admin.yml
```
should be something like:
```
diag
diag
├── kudo
│   └── clusterrole_cluster-admin.yml # fmt.Sprintf(%s_%s), ToLower(Kind), Name)
```

2. The only resources I would nest right now are `Pods` because immediately seeing which `Deployment/Statefulset/Job/...` they belong to might be beneficial:
Example:
```
    │   ├── StatefulSet_zk-zookeeper
    │   │   ├── Pod_zk-zookeeper-0
    │   │   │   └── zk-zookeeper-0.yml
```
should be:
```
    │   ├── statefulset_zk-zookeeper
    │   │   └── pod-zk-zookeeper-0.yml
```
3. Additionally, the above statefulset (and all the other instance resources) should be inside the `operator_zookeeper/operatorversion_zookeeper-0.3.0/instance_zk/` folder.

ANeumann82 commented:
@zen-dog I'm not fully sure about that. I think nested folders for pods may be ok, especially when we add to start additional resources per pod:
- The Pod-Yaml
- The Logs from all containers in the Pod
- Custom Logs/Data that may be specified by the operator (although that may end up somewhere else?)

Apart from that I agree with regards to the folder structure.


Regarding the PR itself: For my taste there's far too much abstraction and indirection - Not a big fan of function refs. In some cases they can be useful, but I had a hard time wrapping my head around the execution flow here. I think instead of having callbacks for `objectGetter` and `objectLister`, an Interface and multiple implementations for different kinds would work better. 

We could even implement generic versions: One that uses the appLabel Selector ( for Pods, Services, Deployments, etc. ), One that doesn't use a selector ( for Events) and then custom ones for the RBAC stuff that actually needs specific handling. The generic versions can use the `KubeClient.DynamicClient` - it returns an `Unstructured`, but that can be converted into a typed object if we really need it.


With regards to building the tree: I think the general approach is correct, it just needs refactoring and cleanup. Instead of `printableTree` as the tree node I would probably use something more like a `directory` structure, with a list of `printable` files and a list of `directory` children.

Maybe it would be an idea to change the attachmentRules code: Instead of having helper functions that map an object to the parent name, use a set of funcs: 
a) return the parent object of an object ( i.e. specific pod -> specific stateful set, specific stateful set -> KUDO instance, etc) 
b) funcs that return the name of a resource (probably r.getName() in most cases)

I'm not fully sure here, I would probably have to play around with the code to really get into it. It is a complex problem :)

nfnt commented on pkg/kudoctl/cmd/diagnostics.go:
```diff
@@ -0,0 +1,53 @@
+func newDiagnosticsCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "diagnostics",
+		Short:   "diagnostics",
```
"Short" is supposed to be a short description of the (sub-)command. Something to consider here and below when polishing this PR.

vemelin-epm commented on pkg/kudoctl/cmd/diagnostics/names.go:
```diff
@@ -0,0 +1,72 @@
+	extractor nameExtractorFn
+}
+
+type nameProviders struct {
```
I may have unnecessarily wrapped some maps, fair point.
I was writing code by first implementing the code that accepts the functionality, hence the functions. Then I implemented the functionality with maps.
I'll try to simplify wherever possible.

zen-dog commented on pkg/kudoctl/cmd/diagnostics.go:
```diff
@@ -0,0 +1,53 @@
+	return cmd
+}
+
+func newDiagnosticsAnalyzeCmd() *cobra.Command {
```
Analyzing bundles is not part of this PR. Please remove

zen-dog commented:
> With that I didn't understand the dependency handling through callback functions and the whole NameProviderFor functionality

To give this some context: you need to fetch these resources in a certain order. In the simplest case, you need the `Instance` to get `Instance.Name` to fetch e.g. pods. A similar relationship exists with some other resources (e.g. role-bindings and service-accounts). 

In general, I'd prefer simply adding the collectors in the right order - we only have to do it once as this is not exposed to the user. A general-purpose solution with dependency graphs or even a half-baked one doesn't seem warranted.

vemelin-epm commented:
Thank you for the comments.
I clearly understand now that the code is **unreadable**.
I'll work on that after accomplishing the functionality.

The point of my efforts in this PR was decoupling of everything possible. The most important part was decoupling of 
- rules how to obtain resource names from other collected resources
- output hierarchy
from the process of collection. Managing those in the process of collection was messy. Now those rules can be managed and modified separately from the collection code.
I've implemented a uniform flow hoping to simplify things:
1. prep phase, add resource collectors. if resource depends on some other resource, this resource is added (recursively) (unless already added), and the name extraction callback is added too.
2. collection phase. for each collector:
- resource is collected
- if it should provide some names (there are an extraction callbacks), extractors are called, names stored
3. build a directory tree. we have a resource cache, from where we can extract resources conveniently by names with just a small overhead and forget about the collectors
4. write to files accumulating directories by tree traversal

nfnt commented:
@vemelin-epm & @zen-dog Thanks for the explanation, that matches how I understood this as well. I'm questioning why this approach is needed, but the approach I have in mind might be too naive; wouldn't something equivalent to `kubectl get pods -n $namespace --selector=kudo.dev/instance=$instance`, repeated for the different kinds of resources achieve the same while being much simpler in implementation?

vemelin-epm commented:
**Status update:**
a discussion offline with @zen-dog showed that I misinterpreted the task for the initial iteration. This PR is designed to allow some extras which are arguably not needed at the moment, if ever.
The difference in brief is that the simplified task requires "collect-time" dependencies only on _Instance_, _OperatorVersion_ and _Operator_ , the rest is just filtered by labels, and the output should be "flat" except for the three resources mentioned.
The task for the first iteration is hereby:
1. For  the _Instance_: 
- collect _only_ the resources that are _created by KUDO_ and therefore marked by the instance-specific set of labels
- collect logs
- store the resources and logs in the hierarchy as follows: _Operator_ is the root, _OperatorVersion_ and _Instance_ are attached to the _Operator_,  all the other resources are attached to the _Instance_ as files except for the _Pods_, each pod being stored alongside with its _Logs_ in its own directory attached to _Instance_
2. For the _Kudo Manager_:
 - collect all the resources
 - collect logs
 - store resources in a flat structure as files, except for _Pods_, each pod being stored in its own directory alongside with its _Logs_
_Simplified PR is under way_
Current output:
```
diag
├── kudo
│   ├── pod_kudo-controller-manager-0
│   │   ├── kudo-controller-manager-0.log.gz
│   │   └── kudo-controller-manager-0.yaml
│   ├── serviceaccountlist.yaml
│   ├── servicelist.yaml
│   └── statefulsetlist.yaml
├── operator_zookeeper
│   ├── instance_zk
│   │   ├── pod_zk-zookeeper-0
│   │   │   ├── zk-zookeeper-0.log.gz
│   │   │   └── zk-zookeeper-0.yaml
│   │   ├── pod_zk-zookeeper-1
│   │   │   ├── zk-zookeeper-1.log.gz
│   │   │   └── zk-zookeeper-1.yaml
│   │   ├── pod_zk-zookeeper-2
│   │   │   ├── zk-zookeeper-2.log.gz
│   │   │   └── zk-zookeeper-2.yaml
│   │   ├── servicelist.yaml
│   │   ├── statefulsetlist.yaml
│   │   └── zk.yaml
│   ├── operatorversion_zookeeper-0.3.0
│   │   └── zookeeper-0.3.0.yaml
│   └── zookeeper.yaml
├── settings.yaml
└── version.yaml
```
@zen-dog please confirm the task stated correctly
@nfnt this seems to match your last comment about the simplified approach

vemelin-epm commented:
Closing the PR due to unnecessary flexibility and complexity.
Work on simplified diagnostics will be in progress here https://github.com/kudobuilder/kudo/pull/1506
